impl/sql/mutationstorage/queue.go

```diff
@@ -101,57 +102,70 @@ func (r *Receiver) run(ctx context.Context, last time.Time) {
 	}
 
 	if time.Since(last) > (r.opts.MaxPeriod - r.opts.Period) {
-		r.sendBatch(ctx, 0, r.opts.MaxBatchSize) // We will be overdue for an epoch soon.
+		if _, err := r.sendMultiBatch(ctx, 0, int(r.opts.MaxBatchSize)); err != nil {
+			glog.Errorf("firstTick: sendMultiBatch(): %v", err)
+		}
 	}
 
 	for {
-		var count int32
 		select {
-		case <-r.more:
-			count = r.sendBatch(ctx, 1, r.opts.MaxBatchSize)
 		case <-r.ticker.C:
-			count = r.sendBatch(ctx, 1, r.opts.MaxBatchSize)
+			if _, err := r.sendMultiBatch(ctx, 1, int(r.opts.MaxBatchSize)); err != nil {
+				glog.Errorf("minTick: sendMultiBatch(): %v", err)
+			}
 		case <-r.maxTicker.C:
-			count = r.sendBatch(ctx, 0, r.opts.MaxBatchSize)
+			if _, err := r.sendMultiBatch(ctx, 0, int(r.opts.MaxBatchSize)); err != nil {
+				glog.Errorf("maxTick: sendMultiBatch(): %v", err)
+			}
 		case <-ctx.Done():
 			return
 		case <-r.done:
 			return
 		}
-		if count >= r.opts.MaxBatchSize {
-			// Continue sending until we drop below batch size.
-			r.more <- time.Now()
+	}
+}
+
+// sendMultiBatch will send multiple batches if the number of available messages > maxMsgs.
+// This helps the queue catch up when there is high traffic, rather than waiting for the next
+// mintick to occur.
+func (r *Receiver) sendMultiBatch(ctx context.Context, minMsgs, maxMsgs int) (int, error) {
+	var total int
+	var err error
+	for sent := maxMsgs; sent >= maxMsgs; {
+		sent, err = r.sendBatch(ctx, minMsgs, maxMsgs)
+		total += sent
+		if err != nil {
+			return total, err
 		}
 	}
+	return total, nil
 }
 
 // sendBatch sends up to batchSize items to the receiver. Returns the number of sent items.
 // If the number of available items is < minBatch, 0 items are sent.
 // If the number of available items is > maxBatch only maxBatch items are sent.
-func (r *Receiver) sendBatch(ctx context.Context, minBatch, maxBatch int32) int32 {
-	ms, err := r.store.readQueue(ctx, r.domainID, maxBatch)
+func (r *Receiver) sendBatch(ctx context.Context, minBatch, maxBatch int) (int, error) {
+	ms, err := r.store.readQueue(ctx, r.domainID, int32(maxBatch))
 	if err != nil {
-		glog.Errorf("readQueue(): %v", err)
-		return 0
+		return 0, fmt.Errorf("readQueue(): %v", err)
 	}
-	if int32(len(ms)) < minBatch {
-		return 0
+	if len(ms) < minBatch {
+		return 0, nil
 	}
 
 	if err := r.recieveFunc(ms); err != nil {
-		glog.Errorf("queue.SendBatch(): %v", err)
-		return 0
+		return 0, fmt.Errorf("queue.SendBatch(): %v", err)
```

Comment:
do you want to note in the error message that it's r.receiveFunc() that returned the non-nil err?

since you're in func queue.sendBatch() wouldn't you normally expect the caller to log a message saying that queue.sendBatch() had errored?

Reply:
yep